Tutorials setup truffle oz and create collectable token (#288)
rsksmart/rsksmart.github.io: tutorials/tokens/create-a-collectable-token.md

---
layout: rsk
title: Create your own collectable token on RSK network
tags: tutorial, rsk, token, openzeppelin, erc721, truffle, frontend, web3, react
description: "How to create your own collectable token on RSK network using Truffle framework, Open Zeppelin libraries and a react frontend."
---

In this tutorial, you will learn about blockchain programming from scratch by building a full decentralized application step-by-step. 

> You will also learn how to create your own collectable token on the RSK blockchain network using the Truffle framework, Open Zeppelin libraries, and build a frontend with the `create-react-app`.

Reviewer: ```suggestion
You will also learn how to create your own collectable token on the RSK blockchain network using the Truffle framework, Open Zeppelin(OZ) libraries, and build a frontend with the `create-react-app`.
```

We will create a dapp inspired by the popular website [Cryptokitties](https://www.cryptokitties.co/), 
a popular blockchain game where you can collect and breed digital cats. 
In this tutorial, instead of collecting felines in our app, we are going to collect exclusive color tokens.

> ### Fungible vs Non-Fungible

Reviewer: ```suggestion
### Fungible vs Non-Fungible Tokens
```

A fungible token represents an asset that can be exchanged for any other asset of equal value in its class.

A currency is an example of a fungible asset. 
A $100 bill is equal to any other $100 bill, 
you can freely exchange these bills with one another because they have the same value, 
no matter the serial number on the specific $100 bill. 
They are fungible bills.

A Non-Fungible Token (NFT) is a unique token, so collectible items are non-Fungibles assets and can be represented by NFTs.

### ERC-721

ERC-721 was the first standard, and currently still the most popular standard, 
for representing non-fungible digital assets. 

The most important properties for this kind of asset is to have a way to check who owns what and a way to move things around.

> It is easy to create new ERC721-compliant contracts by importing it from the OpenZeppelin library and we will do so in this tutorial.

Reviewer: ```suggestion
It is easy to create new ERC721-compliant contracts by importing it from the OZ library and we will do so in this tutorial.
```

Reviewer: For consistency in the naming convention, unless we do away with OZ totally

The interface for ERC-721 provides two methods:

* function `ownerOf`: to know who owns a token
* function `transferFrom`: to transfer ownership of a token

And this is enough to represent an NFT!

### Colors

In this tutorial, We are going to create a non-fungible token to represent our collectible color tokens. 

You will be able to create new color tokens and claim them so that they can be held in a digital blockchain wallet.

# Overview

Here is a summary of the steps to be taken to build our token:

1. Installation requirements;

> 2. Initialize a project using Truffle and Open Zeppelin;

Reviewer: ```suggestion
2. Initialize a project using Truffle and OZ;
```

9. Deploy a smart contract on RSK Testnet using Truffle;
10. Create client side application;
11. Interact with the smart contract.

Steps 1 to 4 are explained in detail in the tutorial links below: 

* [Setup a project with Truffle and OpenZeppelin](/tutorials/ethereum-devs/setup-truffle-oz/)

# Requirements

1. Node.js and NPM (Node Package Manager)
2. Visual Studio Code (VSCode) or any other editor of your choice
3. Truffle
4. Metamask - google chrome extension

The requirements 1 to 3 are explained in detail in the tutorial links below:

* [Setup a project with Truffle and OpenZeppelin](/tutorials/ethereum-devs/setup-truffle-oz/)

For requirement 5, to install Metamask and connect to RSK testnet and get some tR-BTCs are explained in detail in the tutorial link below: 

* [Remix and Metamask with RSK testnet](/tutorials/ethereum-devs/remix-and-metamask-with-rsk-testnet/)

# Setup the project

Create a new folder named `colors`.

Do the steps below, following instructions from the tutorial 
[Setup a project with Truffle and OpenZeppelin](/tutorials/ethereum-devs/setup-truffle-oz/)

1. Initialize an empty Truffle project;
2. Initialize an npm project;

> 3. Install Open Zeppelin;

Reviewer: ```suggestion
3. Install OZ;
```

1. Create React App
2. Web3.js
3. Bootstrap

## Create React App

This is the official template to create single-page React applications. It offers a build setup with no configuration. 

To learn more: [create react app](https://reactjs.org/docs/create-a-new-react-app.html#create-react-app)

In the project folder, at terminal, run:

```
npx create-react-app app --use-npm
```

![create-react-app](/assets/img/tutorials/create-a-collectable-token/image-01.png)

The option `--use-npm` is to select npm  as package manager

[npx](https://medium.com/@maybekatz/introducing-npx-an-npm-package-runner-55f7d4bd282b) comes with npm 5.2+ and higher, see [instructions for older npm versions](https://gist.github.com/gaearon/4064d3c23a77c74a3614c498a8bb1c5f).

This is a large package and this might take a couple of minutes to show the message of successful installation:

![create-react-app successful installation](/assets/img/tutorials/create-a-collectable-token/image-02.png)

Now you have a new folder named `app` and we will customize our frontend later.

![create-react-app file structure](/assets/img/tutorials/create-a-collectable-token/image-03.png)

## Web3.js

> Web3.js helps us to develop websites or clients that interact with the blockchain - writing code that reads and writes data from the blockchain with smart contracts. 

Reviewer: ```suggestion
Web3.js helps us to develop websites or clients that interact with the blockchain - writing code that reads and writes data from the blockchain with smart contracts. 
```

# Smart contract architecture

We will create a smart contract named `Color.sol` that will inherit the ERC721 definition from the OZ library.

## Create the smart contract Color.sol

In the Contracts folder, create a new file named `Color.sol`

![Create Color.sol](/assets/img/tutorials/create-a-collectable-token/image-09.png)

## File Color.sol

Copy and paste the smart contract from the following gist, or inline below:

[Color.sol](https://raw.githubusercontent.com/solangegueiros/tokens-rsk/master/colors-create-react-app/contracts/Color.sol)

```javascript
pragma solidity 0.5.7;

import '@openzeppelin/contracts/token/ERC721/ERC721Full.sol';

contract Color is ERC721Full {

  bytes3[] public colors;
  mapping(bytes3 => bool) private _colorExists;

  constructor() ERC721Full("Color", "COLOR") public {
  }

  // E.G. color = "#FFFFFF"
  function mint(bytes3 _color) public {
    require(!_colorExists[_color], "color exists");
    uint _id = colors.push(_color);
    _mint(msg.sender, _id);
    _colorExists[_color] = true;
  }
}
```

It looked like this:

![Color.sol](/assets/img/tutorials/create-a-collectable-token/image-10.png)

## Understanding the smart contract

> To create our ERC721 Token, we will import `ERC721Full` from Open Zeppelin. 

Reviewer: ```suggestion
To create our ERC-721 Token, we will import `ERC721Full` from OZ. 
```

# Deploy a smart contract at testnet

First of all, we need to create a file in truffle structure with instructions to deploy the smart contract.

## Create file 2_deploy_contracts.js

Folder migrations has JavaScript files that help you deploy contracts to the network. 
These files are responsible for staging your deployment tasks, and they're written under the assumption that your deployment needs will change over time. 
A history of previously run migrations is recorded on-chain through a special Migrations contract, detailed below. 
(source: [running migrations](https://www.trufflesuite.com/docs/truffle/getting-started/running-migrations))

In the migrations folder, create the file `2_deploy_contracts.js`

![create 2_deploy_contracts](/assets/img/tutorials/create-a-collectable-token/image-12.png)

Copy and paste this code:

```javascript
const Color = artifacts.require("Color");

module.exports = function(deployer) {
  deployer.deploy(Color);
};
```

It looked like this:

![2_deploy_contracts](/assets/img/tutorials/create-a-collectable-token/image-13.png)

## Migrate

In the terminal, run this command:

```
truffle migrate --network testnet
```

> Wait a few minutes while the transactions for the smart contract deployments are sent to the blockchain…

Reviewer: ```suggestion
Wait a few minutes while the transactions for the smart contract deployments are sent to the blockchain.
```

                <form onSubmit={(event) => {
                  event.preventDefault();
                  const colorStr = this.color.value;
                  this.mint(colorStr);
                }}>
                  <input
                    type='text'
                    className='form-control mb-1'
                    placeholder='e.g. #FF00FF'
                    ref={(input) => { this.color = input }}
                  />
                  <input
                    type='submit'
                    className='btn btn-block btn-primary'
                    value='MINT'
                  />
                </form>
              </div>
            </main>
          </div>
          <hr/>
          <div className="row text-center">
            { this.state.colors.map((colorStr, key) => {
              return (
                <div key={key} className="col-md-3 mb-3">
                  <div className="token" style={{ backgroundColor: colorStr }}></div>
                  <div>{colorStr}</div>
                </div>
              );
            })}
          </div>
        </div>
      </div>
    );
  }
}

export default App;
```

## Understanding App.js

> After import web3 here:

Reviewer: ```suggestion
Import web3 here:
```

```javascript
import Web3 from 'web3'
```

> This part connected to RSK testnet using the wallet injected, in this case, Metamask:

Reviewer: ```suggestion
This part is connected to the RSK testnet using the web3 wallet injected, in this case, Metamask:
```

  </tr>
  <tr>
    <td>Blue</td>
    <td>#0000FF</td>
  </tr>
  <tr>
    <td>Yellow</td>
    <td>#FFFF00</td>
  </tr>
</table>

## Mint

Choose a color and enter your hexadecimal representation in the info text field, and click on the MINT button.

It will call the `mint()` function at the smart contract instance Color, with the color that you defined.

I will enter the color value `#FF0000`.

> Do not forget to use the symbol `#`

![#FF0000](/assets/img/tutorials/create-a-collectable-token/image-23.png)

![#FF0000 Metamask confirm](/assets/img/tutorials/create-a-collectable-token/image-24.png)

Click at confirm button

Great! Now I have my first color collectable token: 

![Red minted](/assets/img/tutorials/create-a-collectable-token/image-25.png)

I would like to mint the blue color: `#0000FF`

![#0000FF](/assets/img/tutorials/create-a-collectable-token/image-26.png)

Wait for a few seconds for your transaction to be mined…

And now I have two colors in my collection!

![Blue minted](/assets/img/tutorials/create-a-collectable-token/image-27.png)

And my collection is growing!

![color collection](/assets/img/tutorials/create-a-collectable-token/image-28.png)

# Congratulations!

> Did you think that it would be so easy to create a NFT? 

Reviewer: Can we rephrase this question?

> I showed you how to connect Truffle to the RSK network and deploy your own NFT token using the Open Zeppelin libraries, and that they work on the RSK network.

Reviewer: ```suggestion
I showed you how to connect Truffle to the RSK network and deploy your own NFT token using the Open Zeppelin(OZ) libraries, and that they work on the RSK network.
```
